Progress Report - August 2022 (#332)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-aug-2022/index.md

Time will tell what happens to older Windows versions after future changes to Vulkan and memory management are implemented in the future, but for now, they work for a while longer.

Good news doesn't end there, here are a few intended to help the modding community..

In the past, in-line comments added to the pchtxt mods weren’t allowed, since we parsed the text in a way that made them incompatible.
Now Morph {{< gh-hovercard "8755" "delimits the hexadecimal value string," >}} allowing modders to add comments.
For example:

```
Before: here             to here
         v                 v
12345678 12345678 // comment
         ^      ^
After:  here  to here
```

Our current file system emulation is spaghetti code caused by old guesses made when less information was available, and patchwork done by developers that are no longer contributing to the project. 
There’s no way to sugar-coat it, this is the main reason Morph has been working so hard for over a year on `Project Gaia`, a full rewrite of the current virtual file system implementation.

> Meanwhile, some patchwork can be done to ease the head pain on developers and modders alike.
> For example, `Build IDs` are padded by a bunch of zeroes, and modders are more used to the short form without including the irrelevant zeroes at the beginning.
> To alleviate this, byte[] {{< gh-hovercard "8790" "parses the Build ID string using fmt" >}} to dynamically add the required zeroes when needed.

Reviewer: ```suggestion
Meanwhile, some further patchwork can be done to reduce the headaches developers and modders currently endure.
For example, `Build IDs` are padded by a bunch of zeroes, and modders are more used to the short form without including the irrelevant zeroes at the beginning.
To remedy this, byte[] {{< gh-hovercard "8790" "parses the Build ID string using fmt" >}} to dynamically add the required zeroes when needed.
```

## UI changes

People have been asking for stronger audio volume in yuzu for a long time, and Maide delivers.
Now, if the user goes to `Emulation > Configure… > Audio`, they can {{< gh-hovercard "8756" "set the volume up to 200%," >}} really cranking it to 11!

{{< imgs
	"./audio.png| Don't wake up your neighbours"
  >}}

Something to keep in mind is that some games have their volume tuned up to the output they are using.
A game may be louder in handheld over docked, for example, so don’t stress your poor ears.

Part of yuzu’s code is still based on [Citra](https://citra-emu.org/), even some sections that are no longer relevant, or have become outdated over the years.
Cleaning up code is a boring but necessary part of any project, more so for open source ones, and yuzu is no exception.
One feature that was very useful in the early days is loading ELF binaries.
Instead of dealing with XCI and NSP dumps, back in the early days, devs dumped and extracted the whole file structure of the game, and loaded the ELF binary directly, in a similar way to how on Windows the user opens a .exe file to load a program.

> Some years down the line and not only this feature [tarnished out of grace](https://www.youtube.com/watch?v=6R3ouGNcACQ), but it also stopped working.
> Since it’s not a priority at the moment, and it can lead to user confusion, byte[] {{< gh-hovercard "8701" "removed it from the File submenu." >}}

Reviewer: ```suggestion
Some years down the line and not only is this feature [tarnished out of grace](https://www.youtube.com/watch?v=6R3ouGNcACQ), but it also stopped working.
Since it’s not a priority at the moment, and it can lead to user confusion, byte[] {{< gh-hovercard "8701" "removed it from the File dropdown." >}}
```

> Asian Windows users noticed that the Size column in the game list and the Speed Percent setting displayed *very weird* numbers.

Reviewer: ```suggestion
Asian Language Windows users noticed that the Size column in the game list and the Speed Percent setting displayed *very weird* characters.
```

{{< single-title-imgs
    "Alien numbers"
    "./bad1.png"
    "./bad2.png"
    >}}

> Of course this isn’t acceptable, so [Docteh](https://github.com/Docteh) implemented the {{< gh-hovercard "8715" "necessary changes" >}} to workaround this issue under Windows.

Reviewer: ```suggestion
Of course this isn’t acceptable, so [Docteh](https://github.com/Docteh) implemented the {{< gh-hovercard "8715" "necessary changes" >}} to workaround this issue in Windows.
```

{{< imgs
	"./digits.png| Much better"
  >}}

> Following up with Asian only UI bugs, If the user started yuzu for example in Chinese and then switched to English, the Filter bar at the bottom and the game list columns at the top would remain untranslated, refusing to switch.

Reviewer: ```suggestion
Adding to the list of Asian-language-only UI bugs, if a user started yuzu, for example, in Chinese and then switched to English, the Filter bar at the bottom and the game list columns at the top would remain untranslated, refusing to switch.
```

{{< gh-hovercard "8797" "Forcing Qt to retranslate" >}} these objects solves the issue, making Docteh happy.

> Docteh also implemented some {{< gh-hovercard "8741" "nice cleanups" >}} to the About dialog.

Reviewer: ```suggestion
Docteh implemented {{< gh-hovercard "8741" "a couple more UI fixes" >}} as well, cleaning up the About dialog and making the default Port number in the Direct Connect UI ignore translations.
```

{{< single-title-imgs-compare

> 	" "

Reviewer: ```suggestion
	"yuzu About dialog changes"
```

	"./aboutbug.png"
	"./aboutfix.png"
>}}

> Users complained that the Control profiles drop down list had a too short character limit, not allowing them to be creative.
> german77 stepped up and {{< gh-hovercard "8783" "increased the character limit." >}}

Reviewer: ```suggestion
Users complained that the Controller Profile drop down list's character limit was too short, inhibiting their creativity.
german77 stepped up and {{< gh-hovercard "8783" "increased the character limit." >}}
```

## Input improvements

> UI navigation with a controller is relatively recent in yuzu, so there is room for improvement.
> An example reported by users is that if asking for the user profile at boot is enabled, the controller will be able to navigate the options and start the game, but it won’t be able to actually change the profile, or cancel the boot process.
> Newcomer [jonesyUK](https://github.com/jonesyUK) {{< gh-hovercard "8716" "solves both issues" >}} with a bit of Qt magic.
> Thank you!

Reviewer: ```suggestion
UI navigation with a controller is a relatively recent addition to yuzu, so there is room for improvement.
For example, users reported that if the "Prompt for user on game boot" feature is enabled, controller inputs will be able to navigate the options and start the game, but it won’t be able to actually change the profile or cancel the boot process.
Newcomer [jonesyUK](https://github.com/jonesyUK) {{< gh-hovercard "8716" "solves both issues" >}} with a bit of Qt magic.
Thank you!
```

> Some users prefer to depend on ds4windows instead of yuzu’s native support for game controllers.
> While there is a good reason for that, it’s a simple way to have a single app handle XInput emulation, it can also lead to some interesting issues german77 discovered.
> For example, current versions of ds4windows can’t take two sequent vibration commands, producing a (depending on the person you ask) not very comfortable constant vibration.
> To solve this brrrrrracing issue, german77 {{< gh-hovercard "8722" "added a small delay" >}} so the app doesn’t skip the stop signal.

Reviewer: ```suggestion
Some users prefer to depend on [DS4Windows](https://github.com/Ryochan7/DS4Windows) instead of yuzu’s native support for game controllers.
While there is a good reason for that, it’s a simple way to have a single app handle XInput emulation, it can also lead to some interesting issues german77 discovered.
For example, current versions of DS4Windows can’t handle two subsequent vibration commands, producing (depending on the person you ask) an uncomfortable, constant vibration.
To solve this brrrrrracing issue, german77 {{< gh-hovercard "8722" "added a small delay" >}} so the app doesn’t skip the stop signal.
```

On a more colourful topic, and with the intention of fixing a bug with `Mario Party Superstars`, german77 spent many hours cooking {{< gh-hovercard "8724" "how to properly emulate controller colours," >}} with a side dish of fixing battery level icons too!

{{< imgs
	"./colour.png| Art Attack!"
  >}}

> Go to `Emulation > Configure… > Controls > Advanced` and go to town!

Reviewer: ```suggestion
Navigate to `Emulation > Configure… > Controls > Advanced` and go to town!
```

> And to close up, an oopsie from our side.
> With the initial implementation of IR camera support, the output was saved to storage, every single video frame as an individual image file…
> This led to TONS of files being written to disk, making navigation with the file manager a pain, and eating a lot of free storage. Oops.
> german77 acted quickly and {{< gh-hovercard "8805" "moved the output to a buffer." >}}
> 
> If any user suffered from this and can’t remove the files, open a cmd window/terminal, navigate to the Pictures folder, and run `del IMG_00*.jpg` to wipe them all out in a single blow.

Reviewer: ```suggestion
And to close this section out, we have to own up to an oopsie on our side.
With the initial implementation of IR camera support, the output was saved to storage. Every single video frame as an individual image file…
This led to TONS of files being written to disk, making navigation with the file manager a pain, and eating a lot of free storage. Oops.
german77 acted quickly and {{< gh-hovercard "8805" "moved the output to a buffer." >}}

If any user suffered from this and can’t remove the files, open the Command Prompt or Terminal, navigate to the Pictures folder, and run `del IMG_00*.jpg` to wipe them all out in a single blow.
```

## Hardware section

#### Intel ARC joins the fun

> We got information that early adopters of A380 Intel GPUs had issues running their games using Vulkan.
> Since this hardware is not globally available yet, it’s not easy for our developers and testers to get their hands on, so please [post issues in our GitHub](https://github.com/yuzu-emu/yuzu/issues) including [log files](https://yuzu-emu.org/help/reference/log-files/), or contact us in the [forums](https://community.citra-emu.org/c/yuzu-support/14) or our [Discord server](https://discord.gg/u77vRWY).
> We’ll ask for more information if needed to find the reason for this.
> Hurray to the early adopter tax.

Reviewer: ```suggestion
We received information that early adopters of A380 Intel GPUs had issues running their games using Vulkan.
Since this hardware is not globally available yet, it’s not easy for our developers and testers to get their hands on it, so please [post issues in our GitHub](https://github.com/yuzu-emu/yuzu/issues) including [log files](https://yuzu-emu.org/help/reference/log-files/), or contact us on our [forums](https://community.citra-emu.org/c/yuzu-support/14) or our [Discord server](https://discord.gg/u77vRWY).
We’ll ask for more information, if needed, to find the reason for this.
Hurray for the early adopter woes.
```

#### AMD, making use of OpenGL

> Users of AMD GPUs reported about specific points in `Xenoblade Chronicles 3` always resulting in vertex explosions.
> A workaround for this is to run the game with OpenGL and a 2X resolution multiplier, get past the area, save, and return to Vulkan.
> 
> Why the oddly specific 2X resolution multiplier? For the same reason our performance chart last month had no results for `Xenoblade Chronicles 2` on this new OpenGL driver, it leads to a driver related crash.
> 2X solves this crash and AMD was properly reported of this issue. 
> No news so far on it, but since we got access into their Vanguard beta testing program, we’ll tell you as soon as the fix is out in a public driver.
> The game is not that much slower under OpenGL now, but shader stuttering will always be much lower under Vulkan.
> 
> We recommend more emulator developers to [apply to Vanguard](https://www.amd.com/en/technologies/amd-vanguard-program), as the indirect approach of the built-in driver bug report tool is not ideal for the specific needs of console emulators.

Reviewer: ```suggestion
Users with AMD GPUs reported that specific points in `Xenoblade Chronicles 3` always resulted in vertex explosions.
A workaround for this is to run the game with OpenGL and a 2X resolution multiplier, get past the area, save, and swap to Vulkan.

Why the oddly specific 2X resolution multiplier? For the same reason our performance chart last month had no results for `Xenoblade Chronicles 2` on this new OpenGL driver, it leads to a driver related crash.
2X works around this crash and AMD has been properly alerted to this issue. 
No news from them so far, but since we got access into their Vanguard beta testing program, we’ll tell you as soon as the fix is out in a public driver.
The game is not that much slower under OpenGL now, but shader stuttering will always be significantly better under Vulkan.

We recommend more emulator developers to [apply to Vanguard](https://www.amd.com/en/technologies/amd-vanguard-program), as the indirect approach of the built-in driver bug report tool is not ideal for the specific needs of console emulators.
```

#### NVIDIA making us go compute ourselves

We confirmed that at least a big part of the crashes affecting Maxwell and Pascal GPUs by the current 516.XX series of drivers is caused by some driver change in compute shaders, which resonates with our compute shader based ASTC texture decoder.
As a solution, and to avoid this problem in the future as it isn’t the first time an NVIDIA driver update broke it, we’re considering {{< gh-hovercard "8849" "switching to a multithreaded CPU decoder" >}} developed by Morph.

The cause for the shadow flickering on KOEI games is still under investigation.

## Future projects

> The delays in merging `Project Y.F.C.` don’t only affect Mainline users, they also impede progress in development.
> In another example of Hell freezing over (no one expected AMD fixing OpenGL), NVIDIA of all people [open sourced some documentation](https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/3d/clb197.h) (partially of course, this is NVIDIA we’re talking about, they’re so closed source they don’t even get fresh air) including proper names for Maxwell 3D registers, completely new registers and even previously unknown values.
> Maide has been {{< gh-hovercard "8766" "checking, replacing and cleaning a lot of code" >}} a lot of code thanks to this new official information, but the merge for this change will have to wait.
> No game fixes are expected out of this, but the clean-up and ease of understanding is very much worth it.

Reviewer: ```suggestion
The delays in merging `Project Y.F.C.` don’t only affect Mainline users, they also impede progress in development.
In another example of Hell freezing over (no one expected AMD fixing OpenGL), NVIDIA of all people [open sourced some documentation](https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/3d/clb197.h) (partially of course, this is NVIDIA we’re talking about, they’re so closed source they don’t even get fresh air). This includes proper names for Maxwell 3D registers, completely new registers, and even previously unknown values.
Maide has been {{< gh-hovercard "8766" "checking, replacing, and cleaning a lot of code" >}} thanks to this new official information, but the merge for this change will have to wait.
No game fixes are expected out of this, but the clean-up and ease of understanding is very much worth it.
```

> An interesting experiment that will benefit all users in a few bandwidth starved games like `The Legend of Zelda: Breath of the Wild`, `Xenoblade Chronicles 3` and `SHIN MEGAMI TENSEI V` among others is increasing the size of Vulkan’s stream buffer size.
> Testing shows a performance boost of 10-50% depending on the system, even without ReBAR support, which limits the buffer size to 256MB.
> Having access to all available VRAM opens the possibility for even higher gains!
> 
> {{< single-title-imgs-compare
> 	"There should be a difficulty mode called upside-down (Axiom Verge)"
> 	"./128mb.png"
> 	"./2048mb.png"
> >}}

Reviewer: ```suggestion
An interesting experiment that will benefit all users in a few bandwidth-starved games like `The Legend of Zelda: Breath of the Wild`, `Xenoblade Chronicles 3`, and `SHIN MEGAMI TENSEI V`, among others, is increasing the size of Vulkan’s stream buffer size.
Testing shows a performance boost of 10-50% depending on the system, even without ReBAR support, which limits the buffer size to 256MB.
Having access to all available VRAM opens the possibility for even higher gains!

{{< single-title-imgs-compare
	"Look at the performance difference! (Xenoblade Chronicles 3)"
	"./128mb.png"
	"./2048mb.png"
>}}
```

> After too much nagging from your writer, byte[] implemented a {{< gh-hovercard "8861" "draft to easily test the feature with a slider," >}} but the final goal is to manage it dynamically, so don’t expect to see this feature as an UI option, more like just a passive skill.

Reviewer: ```suggestion
After too much nagging from your writer, byte[] implemented a {{< gh-hovercard "8861" "draft to easily test the feature with a slider," >}} but the final goal is to manage it dynamically, so don’t expect to see this feature as a UI option in an official release.
```

Working on the finishing touches for `Project London`, Tobi has not been passive at all:

- {{< gh-hovercard "8685" "Adding the ability to proxy socket packets," >}} which includes custom rooms with local network support, including LAN games, not only LDN.
- {{< gh-hovercard "8731" "Stubbing and adding some required data types." >}}
- {{< gh-hovercard "8822" "Improvements to the in-progress room feature." >}}

{{< imgs

> 	"./ldn.png| The best fun ever doing internal testing"

Reviewer: ```suggestion
	"./ldn.png| The most fun ever doing internal testing"
```

  >}}

> ED: `Project London`, LDN support is out now! More information in [this dedicated article here](https://yuzu-emu.org/entry/ldn-is-here/), and [Multiplayer guide here](https://yuzu-emu.org/help/feature/multiplayer/).
> We’ll cover more in the next report.
> 
> That’s all folks! Thank you so much for your attention until the end, and I hope to see you all next time!
> 
> Special thanks to Mysterious Writer B for your help.

Reviewer: ```suggestion
Oh! How could we almost forget? `Project London`, LDN support, is _finally_ out now! More information in [the dedicated article](https://yuzu-emu.org/entry/ldn-is-here/) and [Multiplayer Guide](https://yuzu-emu.org/help/feature/multiplayer/).
We’ll recap the release in the next Progress Report.

That’s all folks! Thank you so much for sticking with us, and I hope to see you all next time!

Special thanks to Mysterious Writer B for your help on this article!
```

Because yuzu tries to give you as stutter-free an experience as it can, it tries to load all known pipelines into memory before you even start playing. 
This might be a reasonable proposition if a game had just a few thousand pipelines. 
However, keeping Xenoblade's 15000+ distinct pipelines all in memory simultaneously causes the memory usage in the GPU driver to skyrocket.
[Cemu](https://github.com/cemu-project/Cemu) users of old may remember facing a similar issue when using NVIDIA GPUs.

To this end, byte[] looked for any issues in yuzu that might be causing the GPU driver to hold on to more memory than it needs to. 
byte[] noticed that OpenGL tracks individual objects for shader source code in addition to just for compiled shaders. 
Properly {{< gh-hovercard "8682" "allowing these shader source code objects to be freed" >}} reclaims a substantial amount of wasted memory.
Users can expect much lower VRAM and RAM usage now while running OpenGL and [suffering emotional damage](https://www.youtube.com/watch?v=i1ojUmdF42U).

{{< imgs
	"./ears.png| That's a big snuffin' difference (Xenoblade Chronicles 3)"
  >}}

> One advantage yuzu's OpenGL backend has had for `Xenoblade Chronicles 3` for some time was stability. 
> It didn't really crash. 
> Even after the initial Vulkan shader fixes previously mentioned, the game would still randomly crash during cutscenes. 

Reviewer: ```suggestion
One advantage yuzu's OpenGL backend has had for `Xenoblade Chronicles 3` for some time was stability. 
It didn't really crash. However, Vulkan was another story.
Even after the initial Vulkan shader fixes previously mentioned, the game would still randomly crash during cutscenes. 
```
